### DISCUSSION TITLE
[graphcache] Entity resolvers vs list resolvers

### DISCUSSION THREAD

**Discussion by orbiteleven** [urql-graphql/urql#2082] [Q&A] [answered]
I'm a bit confused about how Graphcache uses local resolvers. I have the following type:

```graphql
type Post {
  Comments: [Post!]!
  Parent: Post
  content: String!
  createdAt: DateTime!
  id: String!
  parentId: String
  updatedAt: DateTime!
}
```
and graphcache setup:
```typescript

const parseDateField: Resolver = (parent, _args, _cache, info) =>
  parseISO(parent[info.fieldName] as string)

const post: ResolverMap = {
  createdAt: parseDateField,
  updatedAt: parseDateField,
}

const cache = cacheExchange({
    resolvers: {
      post,
    },
  })

```

This works fine for resolving individual `Post` entities, however when I query `posts` I (think) need a separate resolver setup:
```typescript
setPostDates = (post: PostInput): PostOutput => ({
  ...post,
  createdAt: parseISO(post.createdAt),
  updatedAt: parseISO(post.createdAt),
  Comments: post.Comments?.map?.(setPostDates),
})

const posts: Resolver<PostList> = (parent) =>
  parent?.posts?.length ? parent.posts.map(setPostDates) : undefined

const cache = cacheExchange({
    resolvers: {
      post, // from earlier
    },
    Query: {
        posts
      },
  })
```

Shouldn't the normalization of the `Post` type also apply to a query on `posts`, since they also return `__typename: Post`? Am I missing something?

**Comment by orbiteleven** [comment]
@JoviDeCroock @kitten any chance I could get one of you to chime in here with your infinite urql wisdom? 😁🙏

**Comment by JoviDeCroock** [comment, marked as answer]
Hey!

This is a commonly overlooked bit but in the normalization, the entities are resolved so we know that `Query.post(id:1)` points at `Post:1` but we can't assume that `Query.posts` means all the posts, that would be kinda overreaching it from a library perspective.

If you really want to list out all the in-memory posts that is theoretically possible but I _pesonally_ would not advise it with eye on pagination, ... that could come in the future. However let's look at how we _could_ go about listing out all the fields:

```js
const posts = (parent, fieldArgs, cache, info) => {
  const allPosts = cache.inspectFields('Query').filter(x => info.fieldName === 'post')
  return allPosts.map(field => cache.resolve(entityKey, fieldKey))
}

cacheExchange({
    resolvers: {
      Query: {
          posts
        },
      }
  })
```

I haven't tested this at all 😅 but it's based on our `pagination` helpers, I hope this at least gives an indicator and explanation to what you're seeing.

The tldr for this question in particular is: we can't assume that `Query.posts` will return `__typename: Post` it could be edges, unions, ....

**Reply by orbiteleven** [reply]
Heya! Thanks for the response.

I follow you up until you say that you can't assume that `Query.posts` will return `__typename: Post`; doesn't normalization work on an entry-by-entry basis? Such that it doesn't really matter if `Query.posts` returns an array of `__typename=Foo` entries, because then it should use the `Foo` resolver, no? Isn't that (part of) what normalization is?

To elaborate on the problem here, if I have 
```graphql
type Post {
  Comments: [Post!]!
  Parent: Post
  content: String!
  createdAt: DateTime!
  id: String!
  parentId: String
  updatedAt: DateTime!
}

type Group {
  name: String!
  Posts: [Post!]!
}
```

and I query something like
```graphql
groups {
  id
  name
  Posts {
    id
    content
  }
}
```

Then I would need the same `Post` resolver logic for both `Query.posts` and `Query.groups`. If I have another entity that can have posts, I'd need to do it there too.

**Reply by JoviDeCroock** [reply]
If you have never queried `query.posts(first: 10)` we can't assume the first 10 will be `__typename: Todo` as your back-end should tell us that. The returned entity could be an abstract-type, ....

To your example, if you query a list of groups and want to query `group(id: 1)` we'll need your help ensuring us that `group(id:1)` is `Group:1` as field names have no explicit relation to their type without contacting the back-end

**Reply by orbiteleven** [reply]
Perhaps this is my misunderstanding. Don't object fields require (if we've not defined a custom key) an `id` and a `__typename`? And aren't those used to determined how to normalize the response using either the default or a custom resolver? For example:
```jsonc
{
  // This normalizes using a "group" resolver
  "__typename": "Group",
  "id": "1",
  "name": "group 1",
  "Posts": [
    // This normalizes using a "post" resolver
    {
      "__typename": "Post",
      "id": 1,
      "content": "post 1"
    }
  ]
}
```

Thanks for walking me through this!

**Reply by JoviDeCroock** [reply]
Okay, let's take a step back and consider your original case before we dive into sub-entity resolvers, when we have a list-query:

```gql
query {
  posts { id }
}
```

We can't predict that `posts` will be an array of `__typename: Post`, we can say that _after_ we've gotten the initial response. If we'd get the details for a post with the following query:

```gql
query { post (id: x)
```

Then we can't predict that your schema returns a `Post` so we expect you to tell us that with a resolver or a network call.

When we look at how resolvers are structured we can easily define it as `__typename.fieldName` so in this case: `Query.group` and `Group.posts` and `Post.id|content`. We indeed key by default for `__typename:id`, the point I'm alluding to is the predictability of server-side responses.

To illustrate how the cache would look for the above example:

```
Query.group(id: 1) --> 'Group:1'
Group:1.id --> '1'
Group:1.name: 'group 1'
Group:1.posts --> ['__Post:1']
Post:1.id --> 1
Post:1.content --> 'post 1'
```

So every time we see you hitting the `Query.group` resolver with the argument `id: 1` we can return `Group:1` and look for the fields you are asking for.

We can't however assume that `Query.groups` will return `Group:1` unless a resolver or your server tells us that.

**Reply by orbiteleven** [reply]
Ah, I get it now. It's not about the data we get from the server but rather what *could* come from cache instead.

Thanks again!
